### PR TITLE
Support Combined indexing for __getitem__ and __setitem__

### PR DESCRIPTION
**Pull request by zoooo0820**
### PR types

New features
### PR changes

APIs
### Description

Pcard-66985

#### Background 
At present, the indexing and assignment (i.e. `__getitem__ / __setitem__`)  of Paddle are incomplete in combined indexing (basic-indexing and advanced-indexing both appeared).  This brings great trouble to build some complex models.

- Problem 1: For `__getitem__ ` ,  both dynamic and static mode are not supported combined indexing.
```python
>>> x = paddle.randn((2,3,4))
>>> x[:,[0,1]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/dist-packages/paddle/fluid/dygraph/tensor_patch_methods.py", line 747, in __getitem__
    return self._getitem_index_not_tensor(item)
ValueError: (InvalidArgument) When index contains a list, its length is excepted to 1, but received 2
  [Hint: Expected size == 1, but received size:2 != 1:1.] (at /paddle/Paddle/paddle/fluid/pybind/slice_utils.h:249)
```

- Problem 2: For `__setitem__`, combined indexing forward but not backward is supported in dynamic mode;  both forward and backward are not supported in static mode.
```python
>>> x = paddle.randn((2,2))
>>> x.stop_gradient=False
>>> y = x + 1
>>> y[:,[[0]]] = 10
>>> paddle.grad(y, x)
[Tensor(shape=[2, 2], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       [[1., 1.],
        [1., 1.]])]

# the correct result should be 
[Tensor(shape=[2, 2], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       [[0., 1.],
        [0., 1.]])]
```
```python
>>> paddle.enable_static()
>>> x = paddle.randn((2,2))
>>> x[:, [[0]]] = 10
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/dist-packages/paddle/fluid/framework.py", line 2387, in __setitem__
    return _setitem_impl_(self, item, value)
  File "/usr/local/lib/python3.7/dist-packages/paddle/fluid/variable_index.py", line 684, in _setitem_impl_
    .format(item))
IndexError: Valid index accept int or slice or ellipsis or list, but received [slice(None, None, None), [[0]]].
```

#### What this PR did

1. [New Feature] Support combined indexing for both `__getitem__` and `__setitem__`, with autograd;
2. [Bug fixes] Fix wrong output shape when slices containing a stride in dynamic mode `__getitem__`; e.g. `If x.shape=(2,3,4), x[0, ::2, 1].shape will be [2], while currently outputs [1,2,1], incorrectly `
3. [Optimize] Change the strategy for advanced-indexing in `__setitem__`, using `index_put` instead of  earlier `gather_nd -> elementwise_sub -> scatter_nd_add` or `reshape-> elementwise_mul / reduce_sum -> stack -> scatter -> reshape`.
4. Modify some unit tests since these case are supported and will not raise error in the future.

### CODE REVIEW

File: python/paddle/fluid/variable_index.py [hunk 1]
```diff
@@ -880,3 +881,538 @@ def idx_is_empty(var):
...
+        decrease_axes,
+        advanced_index,
+        has_advanced_index,
+        use_strided_slice,
+    )
+
+
+def _setitem_static(x, indices, values):
+    """
+    [WIP]: support __setitem__ by iteration strategy. combined indexing will be support by this.
```
Comment (jeff41404) [inline python/paddle/fluid/variable_index.py]: Is this WIP compliant with the current situation?

File: python/paddle/fluid/variable_index.py [hunk 2]
```diff
@@ -880,3 +881,538 @@ def idx_is_empty(var):
...
+            none_axes[idx] = new_axis
+
+        out = paddle.unsqueeze(out, axis=none_axes)
+
+    return out
+
+
+def _getitem_static(x, indices):
+    """
+    [WIP]: support __getitem__ by iteration strategy. combined indexing will be support by this.
```
Comment (jeff41404) [inline python/paddle/fluid/variable_index.py]: Is this WIP compliant with the current situation?

Comment (zoooo0820) [reply python/paddle/fluid/variable_index.py]: Done, it is deleted, thanks.

**Review by jeff41404 — DISMISSED** [review]
LGTM

**Review by jeff41404 — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**